### PR TITLE
Add a test for CUDA library build rules

### PR DESCRIPTION
**Pull request by fwyzard**
#### PR description:

This PR adds new packages to test the implementation and usage of CUDA libraries for device code in CMSSW plugins and executables.

The package `HeterogeneousTest/CUDADevice` implements a library that defines and exports CUDA device-only functions, and a plugin and test that use them.

The package `HeterogeneousTest/CUDAKernel` implements a library that imports device functions from `HeterogeneousTest/CUDADevice` to define and export CUDA kernels, and a plugin and test that use them.

The package `HeterogeneousTest/CUDAWrapper` implements a library that imports kernels from `HeterogeneousTest/CUDAKernel` to define and export host-only wrappers around them, usable by non-CUDA libraries, plugins and applications, and implements a plugin and test that use them.

The package `HeterogeneousTest/CUDAOpaque` implements a library that use the wrappers from `HeterogeneousTest/CUDAKernel` to define and export host-only functions around the whole CUDA section, usable by libraries, plugins and applications that are not CUDA-aware, and implements a plugin and test that use them.

#### PR validation:

The new unit tests compile and pass.

### CODE REVIEW

**Comment by fwyzard** [comment]
@makortel this _almost_ works - it needs only a fix in the build command.

**Comment by fwyzard** [comment]
@smuzaffar at least locally, this fails to link because scram does not pass `-lHeterogeneousCoreCUDATestDeviceLib` to gcc, even if the build file does state
```xml
    <use name="HeterogeneousCore/CUDATestDeviceLib"/>
```

The resulting command is 
```bash
>> Building binary testCudaDeviceAddition
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02768/el8_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/c++ -O2 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++1z -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -msse3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-deprecated-copy -Wno-unused-parameter -Wunused -Wparentheses -Wno-deprecated -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DUSE_CMS_DEPRECATED -DBOOST_DISABLE_ASSERTS -fPIC tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestDeviceLib/test/testCudaDeviceAddition/testDeviceAddition.cu.o tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestDeviceLib/test/testCudaDeviceAddition/testCudaDeviceAddition_cudadlink.o -Wl,-E -Wl,--hash-style=gnu -L/data/user/fwyzard/CMSSW_13_0_X_2023-01-21-1100/biglib/el8_amd64_gcc11 -L/data/user/fwyzard/CMSSW_13_0_X_2023-01-21-1100/lib/el8_amd64_gcc11 -L/cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc11/cms/cmssw-patch/CMSSW_13_0_X_2023-01-21-1100/biglib/el8_amd64_gcc11 -L/cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc11/cms/cmssw-patch/CMSSW_13_0_X_2023-01-21-1100/lib/el8_amd64_gcc11 -L/cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc11/cms/cmssw-patch/CMSSW_13_0_X_2023-01-21-1100/external/el8_amd64_gcc11/lib -L/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02768/el8_amd64_gcc11/cms/cmssw/CMSSW_13_0_X_2023-01-19-2300/lib/el8_amd64_gcc11 -L/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02768/el8_amd64_gcc11/cms/cmssw/CMSSW_13_0_X_2023-01-19-2300/biglib/el8_amd64_gcc11 -L/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02768/el8_amd64_gcc11/external/cuda/11.5.2-66a9473808e7d5863d5bbec0824e2c4a/lib64/stubs -lcudart -lcudadevrt -lnvToolsExt -lcuda -o tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestDeviceLib/test/testCudaDeviceAddition/testCudaDeviceAddition
```
and it fails with
```
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02768/el8_amd64_gcc11/external/gcc/11.2.1-f9b9dfdd886f71cd63f5538223d8f161/bin/../lib/gcc/x86_64-redhat-linux-gnu/11.2.1/../../../../x86_64-redhat-linux-gnu/bin/ld: tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestDeviceLib/test/testCudaDeviceAddition/testCudaDeviceAddition_cudadlink.o: in function `__cudaRegisterLinkedBinary_c84dd51f_17_DeviceAddition_cu_bcd44480':
link.stub:(.text+0x121): undefined reference to `__fatbinwrap_c84dd51f_17_DeviceAddition_cu_bcd44480'
collect2: error: ld returned 1 exit status
```

If I add by by hand `-lHeterogeneousCoreCUDATestDeviceLib` the link is successful, and the test actually runs:
```bash
$ ./tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestDeviceLib/test/testCudaDeviceAddition/testCudaDeviceAddition
===============================================================================
All tests passed (2097152 assertions in 1 test case)
```

**Comment by fwyzard** [comment]
please test

**Comment by smuzaffar** [comment]
@fwyzard , https://github.com/cms-sw/cmsdist/pull/8264 should fix this issue. Problem was that there was a restrictions that only packages with at least one `cxx` files can export libs. In this case `HeterogeneousCore/CUDATestDeviceLib` has only `.cu` files that is why its lib was not exported. https://github.com/cms-sw/cmsdist/pull/8264  now allows packages with any source file (cxx, c, fortran, cuda) should export their libs

**Comment by fwyzard** [comment]
Understood... thanks!

**Comment by fwyzard** [comment]
please test with cms-sw/cmsdist#8264

**Comment by fwyzard** [comment]
please test with https://github.com/cms-sw/cmsdist/pull/8264

**Comment by fwyzard** [comment]
please test with cms-sw/cmsdist#8264

**Comment by fwyzard** [comment]
please test with cms-sw/cmsdist#8264

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test with cms-sw/cmsdist#8264

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
hold

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
@smuzaffar this now fails for me on top of `CMSSW_13_0_X_2023-01-29-1100`:
```
>> Cuda Device Code library tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestKernel/src/HeterogeneousCoreCUDATestKernel/libHeterogeneousCoreCUDATestKernel_nv.a 
Copying tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestKernel/src/HeterogeneousCoreCUDATestKernel/libHeterogeneousCoreCUDATestKernel_nv.a to productstore area:
>> Cuda Device Link tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestKernel/src/HeterogeneousCoreCUDATestKernel/HeterogeneousCoreCUDATestKernel_cudadlink.o 
nvlink error   : Undefined reference to '_ZN3cms8cudatest13add_vectors_dEPKdS2_Pdm' in 'tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestKernel/src/HeterogeneousCoreCUDATestKernel/DeviceAdditionKernel.cu.o' (target: sm_60)
nvlink error   : Undefined reference to '_ZN3cms8cudatest13add_vectors_fEPKfS2_Pfm' in 'tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestKernel/src/HeterogeneousCoreCUDATestKernel/DeviceAdditionKernel.cu.o' (target: sm_60)
gmake: *** [config/SCRAM/GMake/Makefile.rules:1730: tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestKernel/src/HeterogeneousCoreCUDATestKernel/HeterogeneousCoreCUDATestKernel_cudadlink.o] Error 255
```

Did anything change in the CUDA rules ?

**Comment by fwyzard** [comment]
> Did anything change in the CUDA rules ?

`CMSSW_13_0_X_2023-01-29-0000` was still fine.

**Comment by fwyzard** [comment]
Looks like the device-side static libraries are no longer being linked.
Comparing `CMSSW_13_0_X_2023-01-29-0000` and `cmssw-config V07-03-18` with `CMSSW_13_0_X_2023-01-29-1100` and `cmssw-config V07-04-02`:
```diff
 >> Cuda Device Link tmp/el8_amd64_gcc11/src/HeterogeneousCore/CUDATestKernel/plugins/HeterogeneousCoreCUDATestKernelPlugins/HeterogeneousCoreCUDATestKernelPlugins_cudadlink.o
 /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02770/el8_amd64_gcc11/external/cuda/11.5.2-66a9473808e7d5863d5bbec0824e2c4a/bin/nvcc \
   -dlink \
   -L${CMSSW_BASE}/static/el8_amd64_gcc11 \
   -L${CMSSW_RELEASE_BASE}/static/el8_amd64_gcc11 \
-  -lHeterogeneousCoreCUDATestKernel_nv \
-  -lHeterogeneousCoreCUDATestDevice_nv \
   -L${CMSSW_BASE}/biglib/el8_amd64_gcc11 \
   -L${CMSSW_BASE}/lib/el8_amd64_gcc11 \
   -L${CMSSW_RELEASE_BASE}/biglib/el8_amd64_gcc11 \
   -L${CMSSW_RELEASE_BASE}/lib/el8_amd64_gcc11 \
   -L${CMSSW_RELEASE_BASE}/external/el8_amd64_gcc11/lib \
   -L/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02770/el8_amd64_gcc11/external/cuda/11.5.2-66a9473808e7d5863d5bbec0824e2c4a/lib64/stubs \
   -lcudadevrt \
   ...
```

**Comment by fwyzard** [comment]
By the way, would it make more sense to move these new packages from `HeterogeneousCore/` to a new subsystem `HeterogeneousTest/` ?

**Comment by smuzaffar** [comment]
`Tool_DependencyDLINK` here https://github.com/cms-sw/cmssw-config/blob/scramv3/SCRAM/GMake/Makefile.rocm#L7 was overriding `Tool_DependencyDLINK` from https://github.com/cms-sw/cmssw-config/blob/scramv3/SCRAM/GMake/Makefile.cuda#L8 . i will it update and test the change

**Comment by smuzaffar** [comment]
locally you can remove  https://github.com/cms-sw/cmssw-config/blob/scramv3/SCRAM/GMake/Makefile.rocm#L7 ( which is not used by rocm and test

**Comment by fwyzard** [comment]
I've merged locally the changes from `cmssw-config V07-05-01`, and the CUDA libraries compile fine now.

**Comment by makortel** [comment]
> By the way, would it make more sense to move these new packages from `HeterogeneousCore/` to a new subsystem `HeterogeneousTest/` ?

I think it could be a good idea, given that many packages are needed for the test. Should we move also `CUDATest` and `AlpakaTest` there?

**Comment by fwyzard** [comment]
> I think it could be a good idea, given that many packages are needed for the test. Should we move also `CUDATest` and `AlpakaTest` there?

Mhm, that would make sense, yes.

**Comment by fwyzard** [comment]
By the way, since we are talking about shuffling stuff around, I think it would also make sense to move the binaries `cudaComputeCapabilities` and `cudaIsEnabled` from `HeterogeneousCore/CUDAServices` to `HeterogeneousCore/CUDAUtilities` or `HeterogeneousCore/CUDACore`.

**Comment by fwyzard** [comment]
please test with cms-sw/cmsdist#8274

**Comment by fwyzard** [comment]
please test with cms-sw/cms-bot#1928,cms-sw/cmsdist#8274

**Comment by fwyzard** [comment]
please test with cms-sw/cms-bot#1928,cms-sw/cmsdist#8274

**Comment by fwyzard** [comment]
+heterogeneous

File: HeterogeneousTest/CUDAKernel/plugins/BuildFile.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<iftool name="cuda-gcc-support">
+  <library file="*.cc *.cu" name="HeterogeneousCoreCUDATestKernelPlugins">
+    <use name="cuda"/>
+    <use name="DataFormats/StdDictionaries"/>
+    <use name="FWCore/Framework"/>
+    <use name="FWCore/ParameterSet"/>
```
Comment (makortel) [inline HeterogeneousTest/CUDAKernel/plugins/BuildFile.xml:6]:
```suggestion
    <use name="FWCore/ParameterSet"/>
    <use name="FWCore/ServiceRegistry"/>
```

File: HeterogeneousTest/CUDAKernel/plugins/BuildFile.xml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+    <use name="HeterogeneousTest/CUDAKernel"/>
```
Comment (makortel) [inline HeterogeneousTest/CUDAKernel/plugins/BuildFile.xml:7]:
```suggestion
    <use name="HeterogeneousTest/CUDAKernel"/>
    <use name="HeterogeneousCore/CUDAServices"/>
```

File: HeterogeneousTest/CUDAOpaque/plugins/BuildFile.xml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<iftool name="cuda-gcc-support">
+  <library file="*.cc" name="HeterogeneousCoreCUDATestOpaquePlugins">
+    <use name="cuda"/>
+    <use name="DataFormats/StdDictionaries"/>
+    <use name="FWCore/Framework"/>
+    <use name="FWCore/ParameterSet"/>
```
Comment (makortel) [inline HeterogeneousTest/CUDAOpaque/plugins/BuildFile.xml:6]:
```suggestion
    <use name="FWCore/ParameterSet"/>
    <use name="FWCore/ServiceRegistry"/>
    <use name="HeterogeneousCore/CUDAServices"/>
```

File: HeterogeneousTest/CUDAWrapper/plugins/BuildFile.xml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<iftool name="cuda-gcc-support">
+  <library file="*.cc" name="HeterogeneousCoreCUDATestWrapperPlugins">
+    <use name="cuda"/>
+    <use name="DataFormats/StdDictionaries"/>
+    <use name="FWCore/Framework"/>
+    <use name="FWCore/ParameterSet"/>
```
Comment (makortel) [inline HeterogeneousTest/CUDAWrapper/plugins/BuildFile.xml:6]:
```suggestion
    <use name="FWCore/ParameterSet"/>
    <use name="FWCore/ServiceRegistry"/>
```

File: HeterogeneousTest/CUDAWrapper/plugins/BuildFile.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    <use name="HeterogeneousTest/CUDAWrapper"/>
```
Comment (makortel) [inline HeterogeneousTest/CUDAWrapper/plugins/BuildFile.xml:7]:
```suggestion
    <use name="HeterogeneousTest/CUDAWrapper"/>
    <use name="HeterogeneousCore/CUDAServices"/>
```

**Review by makortel — COMMENTED** [review]
I think it would be useful to have a `README.md` somewhere that explains the testing setup (e.g. the purpose of each package).

I'm not sure what package would be best for it though. The "best" I can think of would be `CUDATest`, but that package is not part of these tests.

Or should we think about allowing subsystem-level `README.md` files? @smuzaffar

**Comment by fwyzard** [comment]
> Or should we think about allowing subsystem-level `README.md` files?

I would lean in this direction, expanding on the description in the PR.

**Comment by smuzaffar** [comment]
> Or should we think about allowing subsystem-level `README.md` files? @smuzaffar

@makortel , we already have many package level `README.md` files . I do not think that there is any restrictions on where one can add documentation. A sub-system level `README.md` to describe the general purpose of sub-system and package level  `README.md` to describe the package functionality should be allowed.

File: HeterogeneousTest/CUDADevice/plugins/BuildFile.xml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<iftool name="cuda-gcc-support">
+  <library file="*.cc *.cu" name="HeterogeneousCoreCUDATestDevicePlugins">
+    <use name="cuda"/>
+    <use name="DataFormats/StdDictionaries"/>
```
[the remark this answers is not in the document]
Comment (fwyzard) [reply HeterogeneousTest/CUDADevice/plugins/BuildFile.xml:4]: TBH I do not know - I've always been uncertain whether the ROOT dictionaries should be linked explicitly or not.

If I remove it all tests still pass.

[the remark this answers is not in the document]
Comment (fwyzard) [reply HeterogeneousTest/CUDADevice/plugins/BuildFile.xml:4]: Given that it is never mentioned explicitly in _any_ `BuildFile.xml`, I'd say it's safe to remove it.

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
@makoretel do the README.md files look reasonable ?

**Comment by fwyzard** [comment]
These changes should be good to go from my side, I'll let you sign the PR if you are satisfied with it.

**Comment by makortel** [comment]
> > Or should we think about allowing subsystem-level `README.md` files? @smuzaffar
> 
>  we already have many package level `README.md` files . I do not think that there is any restrictions on where one can add documentation. A sub-system level `README.md` to describe the general purpose of sub-system and package level `README.md` to describe the package functionality should be allowed.

I remember trying to add `HeterogeneousCore/README.md` around the time of the CUDA integration, and having trouble with some of our tools not working well with files (instead of directories) inside the subsystem directory. I tested quickly on recent IBs and got
```
$ git add HeterogeneousCore/README.md
The following paths and/or pathspecs matched paths that exist
outside of your sparse-checkout definition, so will not be
updated in the index:
HeterogeneousCore/README.md
hint: If you intend to update such entries, try one of the following:
hint: * Use the --sparse option.
hint: * Disable or modify the sparsity rules.
hint: Disable this message with "git config advice.updateSparsePath false"
```
although the suggested `git add --sparse HeterogeneousCore/README.md`.

Trying to use the `README.md` in another developer area results e.g.
```
$ git cms-addpkg HeterogeneousCore/README.md

Checking out packages
HeterogeneousCore/README.md

These packages do not exist in branch debug
HeterogeneousCore/README.md
```
Doing  `git cms-addpkg HeterogeneousCore` adds also the `README.md`.

I see scram itself handles that `README.md` file correctly. So ok, I didn't see any real showstoppers in this quick test, but some tools have annoying behavior. I'd bet some other tools would have "strange" behavior with subsystem level files as well.

File: HeterogeneousTest/CUDADevice/README.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+`CUDATestDeviceAdditionModule` plugin.
+
+
+# Other packages
+
+For various ways in which this library and plugin can be tested, see also the other
+`HeterogeneousTest/CUDA...` packages:
+  - `HeterogeneousTest/CUDAKernel/README.md`
+  - `HeterogeneousTest/CUDAWrapper/README.md`
+  - `HeterogeneousTest/CUDAOpaque/README.md`
```
Comment (makortel) [inline HeterogeneousTest/CUDADevice/README.md:40]:
```suggestion
  - [`HeterogeneousTest/CUDAKernel/README.md`](../../HeterogeneousCore/CUDAKernel/README.md)
  - [`HeterogeneousTest/CUDAWrapper/README.md`](../../HeterogeneousCore/CUDAWrapper/README.md)
  - [`HeterogeneousTest/CUDAOpaque/README.md`](../../HeterogeneousCore/CUDAOpaque/README.md)
```

File: HeterogeneousTest/CUDAKernel/README.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+`CUDATestKernelAdditionModule` module.
+
+
+# Other packages
+
+For various ways in which this library and plugin can be tested, see also the other
+`HeterogeneousTest/CUDA...` packages:
+  - `HeterogeneousTest/CUDADevice/README.md`
+  - `HeterogeneousTest/CUDAWrapper/README.md`
+  - `HeterogeneousTest/CUDAOpaque/README.md`
```
Comment (makortel) [inline HeterogeneousTest/CUDAKernel/README.md:40]:
```suggestion
  - [`HeterogeneousTest/CUDADevice/README.md`](../../HeterogeneousTest/CUDADevice/README.md)
  - [`HeterogeneousTest/CUDAWrapper/README.md`](../../HeterogeneousTest/CUDAWrapper/README.md)
  - [`HeterogeneousTest/CUDAOpaque/README.md`](../../HeterogeneousTest/CUDAOpaque/README.md)
```

File: HeterogeneousTest/CUDAOpaque/README.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+`CUDATestOpqaueAdditionModule` module.
+
+
+# Other packages
+
+For various ways in which this library and plugin can be tested, see also the other
+`HeterogeneousTest/CUDA...` packages:
+  - `HeterogeneousTest/CUDADevice/README.md`
+  - `HeterogeneousTest/CUDAKernel/README.md`
+  - `HeterogeneousTest/CUDAWrapper/README.md`
```
Comment (makortel) [inline HeterogeneousTest/CUDAOpaque/README.md:40]:
```suggestion
  - [`HeterogeneousTest/CUDADevice/README.md`](../../HeterogeneousTest/CUDADevice/README.md)
  - [`HeterogeneousTest/CUDAKernel/README.md`](../../HeterogeneousTest/CUDAKernel/README.md)
  - [`HeterogeneousTest/CUDAWrapper/README.md`](../../HeterogeneousTest/CUDAWrapper/README.md)
```

File: HeterogeneousTest/CUDAWrapper/README.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+`CUDATestWrapperAdditionModule` module.
+
+
+# Other packages
+
+For various ways in which this library and plugin can be tested, see also the other
+`HeterogeneousTest/CUDA...` packages:
+  - `HeterogeneousTest/CUDADevice/README.md`
+  - `HeterogeneousTest/CUDAKernel/README.md`
+  - `HeterogeneousTest/CUDAOpaque/README.md`
```
Comment (makortel) [inline HeterogeneousTest/CUDAWrapper/README.md:42]:
```suggestion
  - [`HeterogeneousTest/CUDADevice/README.md`](../../HeterogeneousTest/CUDADevice/README.md)
  - [`HeterogeneousTest/CUDAKernel/README.md`](../../HeterogeneousTest/CUDAKernel/README.md)
  - [`HeterogeneousTest/CUDAOpaque/README.md`](../../HeterogeneousTest/CUDAOpaque/README.md)
```

**Comment by fwyzard** [comment]
> I found compilation error when building:

that's strange - this PR builds for me on machines both with and without GPUs.

**Comment by fwyzard** [comment]
enable gpu

**Review by makortel — COMMENTED** [review]
Thanks @fwyzard, the READMEs look good. I'd only suggest to make the pointers to the READMEs of other packages direct links to those files.

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please abort

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
+heterogeneous

File: HeterogeneousTest/CUDAOpaque/test/testCUDATestAdditionModules.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    size = cms.uint32( 1024*1024 )
+)
+
+process.cudaTestOpaqueAdditionModule = cms.EDAnalyzer('CUDATestOpaqueAdditionModule',
+    size = cms.uint32( 1024*1024 )
+)
+
+process.path = cms.Path(
+    # this one fails with "cudaErrorInvalidDeviceFunction: invalid device function"
+    process.cudaTestDeviceAdditionModule +
```
Comment (makortel) [inline HeterogeneousTest/CUDAOpaque/test/testCUDATestAdditionModules.py:27]: Following up on https://github.com/cms-sw/cmssw/pull/40637/files#r1091756724 on the ROCm side, this test should then fail on GPU_X IB, right?

Comment (fwyzard) [reply HeterogeneousTest/CUDAOpaque/test/testCUDATestAdditionModules.py:27]: yes - ouch, I though I had commented it out :-(

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by rappoccio** [comment]
+1

* Tests are now successful.

**Pull request merged**